File: Kin Unity/Assets/Kin/Editor/NativeCode/iOS/KinPlugin.swift
//
//  KinPlugin.swift
//  Unity-iPhone
//
//  Created by desaro on 11/2/18.
//
import UIKit
import Foundation
import KinSDK
import KinBackupRestoreModule
import Sodium


struct Provider: ServiceProvider {
	public let url: URL
	public let network: Network
	
	static public let Test = Provider( url: URL( string: "http://horizon-testnet.kininfrastructure.com" )!, network: .testNet )
	static public let Main = Provider( url: URL( string: "https://horizon-ecosystem.kininfrastructure.com" )!, network: .mainNet )
	
	
	init( url: URL, network: Network ) {
		self.url = url
		self.network = network
	}
}

enum BackupAction {
    case backup(String)
    case restore(String)
}


@objc class KinPlugin : NSObject {
	@objc static let instance = KinPlugin()
	
	var clients: Dictionary<String,KinClient> = Dictionary<String,KinClient>()
	var accounts: Dictionary<String,KinAccount> = Dictionary<String,KinAccount>()
	var transactions: Dictionary<String,TransactionEnvelope> = Dictionary<String,TransactionEnvelope>()
	
	var paymentListeners = Dictionary<String,PaymentWatch>()
	var balanceListeners = Dictionary<String,BalanceWatch>()
	var accountListeners = Dictionary<String,Promise<Void>>();
	
	let formatter = NumberFormatter()
	var networkId: String = ""
    
    let vc: UIViewController = UnityGetGLViewController()
    
    let backupRestoreManager = KinBackupRestoreManager()
    var lastBackupRestoreRequest: BackupAction?

	override init() {
        super.init()
		formatter.generatesDecimalNumbers = true
        self.backupRestoreManager.delegate = self
	}

	// MARK: - helpers
	
	private func unitySendMessage( method: String, param: String ) {
		UnitySendMessage( "KinManager", method, param )
	}
	
	
	private func errorToJson( error: Error, accountId: String? ) -> String {
		var dict = ["Message": error.localizedDescription, "NativeType": String(describing: error), "ErrorCode": NSNumber.init( value: 1 )] as [String : Any]
		
		if( accountId != nil ) {
			dict["AccountId"] = accountId
		}
		
		do {
			let jsonData = try JSONSerialization.data( withJSONObject: dict, options: [] )
			let jsonString = String( data: jsonData, encoding: .utf8 )
			
			return jsonString!
		}
		catch {
			print( "Error converting error to JSON: \(error)" )
			return "{}"
		}
	}
	
	
	private func callbackToJson( accountId: String, value: String ) -> String {
		let dict = ["AccountId": accountId, "Value": value]
		
		do {
			let jsonData = try JSONSerialization.data( withJSONObject: dict, options: [] )
			let jsonString = String( data: jsonData, encoding: .utf8 )
			
			return jsonString!
		}
		catch {
			return "{}"
		}
	}
	
	
	private func paymentInfoToJson( paymentInfo: PaymentInfo, accountId: String ) -> String {
		let dateFormatter = DateFormatter()
		dateFormatter.dateFormat = "yyyy-MM-dd hh:mm:ss"
		
		var dict = ["AccountId": accountId] as [String : Any]
		dict["_Amount"] = formatter.string( from: paymentInfo.amount as NSDecimalNumber )
		dict["CreatedAt"] = dateFormatter.string( from: paymentInfo.createdAt )
		dict["DestinationPublicKey"] = paymentInfo.destination
		dict["SourcePublicKey"] = paymentInfo.source
		dict["Hash"] = paymentInfo.hash
		dict["Memo"] = paymentInfo.memoText
		
		do {
			let jsonData = try JSONSerialization.data( withJSONObject: dict, options: [] )
			let jsonString = String( data: jsonData, encoding: .utf8 )
			
			return jsonString!
		}
		catch {
			return "{}"
		}
	}
	
	
	private func transactionToJson( transaction: TransactionEnvelope, accountId: String, transactionId: String ) -> String {
		do {
			// do some hacking of the Decodable protocol to get the iOS data to match the Android data
			let whitelistEnvelope = WhitelistEnvelope(transactionEnvelope: transaction, networkId: self.networkId)
			let encodedEnvelope = try? JSONEncoder().encode(whitelistEnvelope)
			let whitelistDict = try! JSONSerialization.jsonObject(with: encodedEnvelope!, options: []) as? [String:String]
			let whitelistPayload = whitelistDict!["tx_envelope"]
			
			var dict = ["AccountId": accountId] as [String : Any]
			dict["Id"] = transactionId
			dict["WhitelistableTransactionPayLoad"] = whitelistPayload
			dict["WhitelistableTransactionNetworkPassphrase"] = Network.testNet.id

			let jsonData = try JSONSerialization.data( withJSONObject: dict, options: [] )
			let jsonString = String( data: jsonData, encoding: .utf8 )

			return jsonString!
		}
		catch {
			return "{}"
		}
	}
	
	
	// MARK: - KinClient
	
	@objc public func createClient( clientId: String, environment: Int, apiKey: String, storeKey: String = "" )
	{
		let provider = environment == 0 ? Provider.Test : Provider.Main
		
		do {
			let appId = try AppId(apiKey)
			let client = KinClient( provider: provider, appId: appId )
			self.clients[clientId] = client
			self.networkId = client.network.id
		}
		catch {
			print( "could not create KinClient: \(error)" )
		}
	}
	
	
	@objc public func getMinimumFee( clientId: String ) {
		if let client = self.clients[clientId] {
			client.minFee().then { (fee) in
				self.unitySendMessage( method: "GetMinimumFeeSucceeded", param: self.callbackToJson( accountId: clientId, value: fee.description ) )
			}.error { (error) in
				print( "getMinimumFee failed: \(error)" )
				self.unitySendMessage( method: "GetMinimumFeeFailed", param: self.errorToJson( error: error, accountId: clientId ) )
			}
		}
	}
	
	
	@objc public func freeCachedClient( clientId: String ) {
		if let _ = self.clients[clientId] {
			print( "Freeing cached client" )
			self.clients.removeValue( forKey: clientId )
		}
	}
	
	
	@objc public func importAccount( clientId: String, accountId: String, exportedJson: String, passphrase: String ) -> String? {
		do {
			if let client = self.clients[clientId] {
				let kinAccount = try client.importAccount(exportedJson, passphrase: passphrase)
				self.accounts[accountId] = kinAccount
				return "";
			}
		}
		catch {
			print( "KinAccount couldn't be imported: \(error)" )
			return errorToJson( error: error, accountId: accountId )
		}
		return nil;
	}
	
	
	@objc public func getAccountCount( clientId: String ) -> Int {
		return self.clients[clientId]?.accounts.count ?? -1
	}
	
	
	@objc public func addAccount( clientId: String, accountId: String ) -> String? {
		if let client = self.clients[clientId] {
			do {
				let account = try client.addAccount()
				self.accounts[accountId] = account
			}
			catch {
				print( "KinAccount couldn't be added: \(error)" )
				return errorToJson( error: error, accountId: accountId )
			}
		}
		return nil
	}
	
	
	@objc public func getAccount( clientId: String, accountId: String, index: Int ) -> Bool {
		if let client = self.clients[clientId] {
			if let account = client.accounts[index] {
				self.accounts[accountId] = account
				return true
			}
		}
		return false
	}
	
	
	@objc public func deleteAccount( clientId: String, index: Int ) -> String? {
		if let client = self.clients[clientId] {
			let account = client.accounts[index]
			
			do {
				// find the key from our dict *before* deleting the account to avoid accessing a deleted account
				var foundKey: String?
				for key in self.accounts.keys {
					let temp = self.accounts[key]
					if temp?.publicAddress == account?.publicAddress {
						foundKey = key
						break
					}
				}
				
				try client.deleteAccount( at: index )

				if foundKey != nil {
					self.accounts.removeValue( forKey: foundKey! )
				}
			}
			catch {
				print( "KinAccount couldn't be deleted: \(error)" )
				return errorToJson( error: error, accountId: nil )
			}
		}
		return nil
	}
	
	
	@objc public func clearAllAccounts( clientId: String ) {
		if let client = self.clients[clientId] {
			client.deleteKeystore()
		}
	}
	
	
	// MARK: - KinAccount
	
	@objc public func freeCachedAccount( accountId: String ) {
		if self.accounts.keys.contains( accountId ) {
			print( "Freeing cached account" )
			self.accounts.removeValue( forKey: accountId )
		}
	}
	
	
	@objc public func getPublicAddress( accountId: String ) -> String {
		return (self.accounts[accountId]?.publicAddress)!
	}
	
	
	@objc public func export( accountId: String, passphrase: String ) -> String? {
		if let account = self.accounts[accountId] {
			do {
				return try account.export( passphrase: passphrase )
			}
			catch {
				print( "KinAccount couldn't be exported: \(error)" )
				return errorToJson( error: error, accountId: nil )
			}
		}
		return nil
	}
	
	
	@objc public func getStatus( accountId: String ) {
		if let account = self.accounts[accountId] {
			account.status { (accountStatus, error) in
				if let error = error {
					print( "getStatus failed: \(error)" )
					self.unitySendMessage( method: "GetStatusFailed", param: self.errorToJson( error: error, accountId: accountId ) )
				} else {
					// normalize to match Android
					var statusInt = accountStatus!.rawValue
					if statusInt == 1 {
						statusInt += 1
					}
					self.unitySendMessage( method: "GetStatusSucceeded", param: self.callbackToJson(accountId: accountId, value: String( statusInt ) ) )
				}
			}
		}
	}
	
	
	@objc public func getBalance( accountId: String ) {
		if let account = self.accounts[accountId] {
			account.balance { (kin, error) in
				if let error = error {
					print( "getBalance failed: \(error)" )
					self.unitySendMessage( method: "GetBalanceFailed", param: self.errorToJson( error: error, accountId: accountId ) )
				} else {
					self.unitySendMessage( method: "GetBalanceSucceeded", param: self.callbackToJson(accountId: accountId, value: (kin?.description)!) )
				}
			}
		}
	}
	
	
	@objc public func buildTransaction( accountId: String, toAddress: String, kinAmount: String, fee: UInt32, memo: String! = nil ) {
		if let account = self.accounts[accountId] {
			account.generateTransaction(to: toAddress, kin: Decimal(string: kinAmount)!, memo: memo, fee: fee) { (transaction, error) in
				if let error = error {
					print( "buildTransaction failed: \(error)" )
					self.unitySendMessage( method: "BuildTransactionFailed", param: self.errorToJson( error: error, accountId: accountId ) )
				} else {
					do {
						let transactionIdHash = try transaction?.tx.hash(networkId: self.networkId)
						let transactionId = transactionIdHash!.hexString
						self.transactions[transactionId] = transaction
						self.unitySendMessage( method: "BuildTransactionSucceeded", param: self.transactionToJson(transaction: transaction!, accountId: accountId, transactionId: transactionId ) )
					} catch {
						print( "buildTransaction failed: \(error)" )
						self.unitySendMessage( method: "BuildTransactionFailed", param: self.errorToJson( error: error, accountId: accountId ) )
					}
				}
			}
		}
	}
	
	
	@objc public func sendTransaction( accountId: String, id: String ) {
		if let account = self.accounts[accountId] {
			if let transaction = self.transactions[id] {
				self.transactions.removeValue(forKey: id)
				account.sendTransaction(transaction) { (transactionId, error) in
					if let error = error {
						print( "sendTransaction failed: \(error)" )
						self.unitySendMessage( method: "SendTransactionFailed", param: self.errorToJson( error: error, accountId: accountId ) )
					} else {
						self.unitySendMessage( method: "SendTransactionSucceeded", param: self.callbackToJson(accountId: accountId, value: transactionId!) )
					}
				}
			}
		}
	}
	
	
	@objc public func sendWhitelistTransaction( accountId: String, id: String, whitelist: String ) {
		if let account = self.accounts[accountId] {
			if self.transactions[id] != nil {
				var envelope: TransactionEnvelope
				do {
					envelope = try XDRDecoder.decode(TransactionEnvelope.self, data: Data(base64Encoded: whitelist)!)
				} catch {
					print( "TransactionEnvelope.decodeResponse failed: \(error)" )
					self.unitySendMessage( method: "SendTransactionFailed", param: self.errorToJson( error: error, accountId: accountId ) )
					return
				}
				
				self.transactions.removeValue(forKey: id)
				account.sendTransaction(envelope) { (transactionId, error) in
					if let error = error {
						print( "sendTransaction failed: \(error)" )
						self.unitySendMessage( method: "SendTransactionFailed", param: self.errorToJson( error: error, accountId: accountId ) )
					} else {
						self.unitySendMessage( method: "SendTransactionSucceeded", param: self.callbackToJson(accountId: accountId, value: transactionId!) )
					}
				}
			}
		}
	}
	
	
	// MARK: - KinAccount Listeners
	
	@objc public func addPaymentListener( accountId: String ) {
		if let account = self.accounts[accountId] {
			let watch = try? account.watchPayments(cursor: nil)
			watch?.emitter.on(queue: .main, next: { [weak self] payment in
				let paymentInfoJson = self?.paymentInfoToJson( paymentInfo: payment, accountId: accountId)
				self?.unitySendMessage( method: "OnPayment", param: paymentInfoJson! )
			})
			
			self.paymentListeners[accountId] = watch
		}
	}
	
	
	@objc public func removePaymentListener( accountId: String ) {
		if self.paymentListeners.keys.contains( accountId ) {
			self.paymentListeners.removeValue( forKey: accountId )
		}
	}
	
	
	@objc public func addBalanceListener( accountId: String ) {
		if let account = self.accounts[accountId] {
			let watch = try? account.watchBalance(nil)
			watch?.emitter.on(queue: .main, next: { [weak self] balance in
				let callbackJson = self?.callbackToJson(accountId: accountId, value: balance.description)
				self?.unitySendMessage( method: "OnBalance", param: callbackJson! )
			})
			
			self.balanceListeners[accountId] = watch
		}
	}
	
	
	@objc public func removeBalanceListener( accountId: String ) {
		if self.balanceListeners.keys.contains( accountId ) {
			self.balanceListeners.removeValue( forKey: accountId )
		}
	}
	
	
	@objc public func addAccountCreationListener( accountId: String ) {
		if let account = self.accounts[accountId] {
			let watch = try? account.watchCreation()
			watch?.finally {
				self.unitySendMessage( method: "OnAccountCreated", param: accountId )
			}
			
			self.accountListeners[accountId] = watch
		}
	}
	
	
	@objc public func removeAccountCreationListener( accountId: String ) {
		if self.accountListeners.keys.contains( accountId ) {
			self.accountListeners.removeValue( forKey: accountId )
		}
	}
    
    @objc public func restoreAccount( clientId: String ) {
        if let client = self.clients[clientId] {
            lastBackupRestoreRequest = BackupAction.restore(clientId);
            backupRestoreManager.restore(client, presentedOnto: self.vc)
        }
    }
    
    @objc public func backupAccount(accountId: String) {
        if let account = accounts[accountId] {
            lastBackupRestoreRequest = BackupAction.backup(accountId);
            backupRestoreManager.backup(account, presentedOnto: vc)
        }
    }
    
    private func generateUniqueId() -> String {
        // Generate a unique id (11 lowercase/numbers), equvilent of the android/c# methods in the Unity relevant code
        return String(NSUUID().uuidString.prefix(11))
    }
}

extension KinPlugin: KinBackupRestoreManagerDelegate {
    func kinBackupRestoreManagerDidComplete(_ manager: KinBackupRestoreManager, kinAccount: KinAccount?) {
        guard let request = lastBackupRestoreRequest else {
            return
        }
        
        switch request {
        case .backup(let accountId):
            unitySendMessage(method: "BackupSucceeded", param: callbackToJson(accountId: accountId, value: ""))
        case .restore(let clientId):
            guard let kAccount = kinAccount else {
                return
            }
            
            let accountId: String

            if let matchedId = accounts
                .first(where: { $1.publicAddress == kAccount.publicAddress})?
                .key {
                accountId = matchedId
            } else {
                accountId = generateUniqueId()
                accounts[accountId] = kAccount
            }

            unitySendMessage(method: "RestoreSucceeded",
                             param: callbackToJson(accountId: clientId, value: accountId))
                
        }
    }
    
    func kinBackupRestoreManagerDidCancel(_ manager: KinBackupRestoreManager) {
        guard let request = lastBackupRestoreRequest else {
            return
        }
        
        switch request {
        case .backup(let accountId):
            unitySendMessage(method: "BackupCanceled", param: self.callbackToJson(accountId: accountId, value: ""))
        case .restore(let clientId):
            unitySendMessage(method: "RestoreCanceled", param: self.callbackToJson(accountId: clientId, value: ""))
        }
    }
    
    func kinBackupRestoreManager(_ manager: KinBackupRestoreManager, error: Error) {
        guard let request = lastBackupRestoreRequest else {
            return
        }
        
        switch request {
        case .backup(let accountId):
            unitySendMessage(method: "BackupFailed", param:
                errorToJson(error: error, accountId: accountId))
        case .restore(let clientId):
            unitySendMessage(method: "RestoreFailed", param:
                errorToJson(error: error, accountId: clientId))
        }
    }
}

File: Kin Android/app/src/main/java/kin/unity/KinPlugin.java
package kin.unity;


import android.text.TextUtils;
import android.util.Log;

import java.math.BigDecimal;
import java.util.HashMap;

import kin.sdk.Balance;
import kin.sdk.KinAccount;
import kin.sdk.KinClient;
import kin.sdk.ListenerRegistration;
import kin.sdk.Transaction;
import kin.sdk.TransactionId;
import kin.sdk.exception.DeleteAccountException;
import kin.sdk.exception.CryptoException;
import kin.sdk.Environment;


@SuppressWarnings("unused")
public class KinPlugin extends KinPluginBase
{
	private static final KinPlugin _instance = new KinPlugin();

	// cache for our native classes that are mirrored in C#
	private HashMap<String,KinClient> _clients = new HashMap<>();
	private HashMap<String,KinAccount> _accounts = new HashMap<>();
	private HashMap<String,Transaction> _transactions = new HashMap<>();

	// ListenerRegistration cache
	private HashMap<String,ListenerRegistration> _paymentListeners = new HashMap<>();
	private HashMap<String,ListenerRegistration> _balanceListeners = new HashMap<>();
	private HashMap<String,ListenerRegistration> _accountListeners = new HashMap<>();


	public static KinPlugin instance()
	{
		return _instance;
	}


	//region KinClient

	public void createClient( String clientId, int environment, String appId, String storeKey )
	{
		// KinClient defaults to null so we do as well here
		if( storeKey == null )
			storeKey = "";

		Environment env = environment == 0 ? Environment.TEST : Environment.PRODUCTION;
		KinClient client = new KinClient( getActivity(), env, appId, storeKey );
		_clients.put( clientId, client );
	}


	public KinClient _getClient( String clientId )
	{
		return _clients.get( clientId );
	}


	public void freeCachedClient( String clientId )
	{
		if( _clients.containsKey( clientId ) )
		{
			Log.i( TAG, "freeing cached client: " + clientId );
			_clients.remove( clientId );
		}
	}


	public String importAccount( String clientId, String accountId, String exportedJson, String passphrase )
	{
		try
		{
			KinAccount account = _clients.get( clientId ).importAccount( exportedJson, passphrase );
			_accounts.put( accountId, account );
			return "";
		}
		catch( Exception e )
		{
			e.printStackTrace();
			return exceptionToJson( e, null );
		}
	}


	public int getAccountCount( String clientId )
	{
		return _clients.get( clientId ).getAccountCount();
	}


	public String addAccount( String clientId, String accountId )
	{
		try
		{
			Log.i( TAG, "adding account: " + accountId );
			KinAccount account = _clients.get( clientId ).addAccount();
			Log.i( TAG, "added account successfully" );
			_accounts.put( accountId, account );
		}
		catch( Exception e )
		{
			e.printStackTrace();
			return exceptionToJson( e, null );
		}

		return "";
	}


	public boolean getAccount( String clientId, String accountId, int index )
	{
		KinAccount account = _clients.get( clientId ).getAccount( index );
		if( account == null )
			return false;

		if( !_accounts.containsValue( account ) )
			_accounts.put( accountId, account );

		return true;
	}


	public KinAccount _getRawAccount( String clientId, int index )
	{
		return _clients.get( clientId ).getAccount( index );
	}


	public String deleteAccount( String clientId, int index )
	{
		Log.i( TAG, "deleting account: " + index );
		KinAccount account = _clients.get( clientId ).getAccount( index );
		if( account == null )
		{
			Log.i( TAG, "could not find account to delete" );
			String message = "Attempted to delete account that doesn't exist at index " + index;
			Log.e( TAG, message );
			return exceptionToJson( new IndexOutOfBoundsException( message ), null );
		}

		try
		{
			// attempt to delete the account
			_clients.get( clientId ).deleteAccount( index );
		}
		catch( Exception e )
		{
			e.printStackTrace();
			return exceptionToJson( e, null );
		}

		return "";
	}


	public void clearAllAccounts( String clientId )
	{
		try
		{
			_clients.get( clientId ).clearAllAccounts();
		}
		catch( Exception e )
		{
			e.printStackTrace();
		}
	}

	//endregion


	//region KinAccount

	public void freeCachedAccount( final String accountId )
	{
		if( _accounts.containsKey( accountId ) )
		{
			Log.i( TAG, "freeing cached account: " + accountId );
			_accounts.remove( accountId );
		}
	}


	public String getPublicAddress( final String accountId )
	{
		return _accounts.get( accountId ).getPublicAddress();
	}


	public String export( final String accountId, final String passphrase )
	{
		try
		{
			return _accounts.get( accountId ).export( passphrase );
		}
		catch( Exception e )
		{
			e.printStackTrace();
			return exceptionToJson( e, null );
		}
	}


	public void getStatus( final String accountId )
	{
		new Thread( () -> {
			try
			{
				int status = _accounts.get( accountId ).getStatusSync();
				unitySendMessage( "GetStatusSucceeded", callbackToJson( String.valueOf( status ), accountId ) );
			}
			catch( Exception e )
			{
				unitySendMessage( "GetStatusFailed", exceptionToJson( e, accountId ) );
				Log.e( TAG, "GetStatus failed", e );
			}
		} ).start();
	}


	public void getBalance( final String accountId )
	{
		new Thread( () -> {
			try
			{
				Balance balance = _accounts.get( accountId ).getBalanceSync();
				unitySendMessage( "GetBalanceSucceeded", callbackToJson( balance.value().toString(), accountId ) );
			}
			catch( Exception e )
			{
				unitySendMessage( "GetBalanceFailed", exceptionToJson( e, accountId ) );
				Log.e( TAG, "GetBalance failed", e );
			}
		} ).start();
	}


	public void getMinimumFee( final String clientId )
	{
		new Thread( () -> {
			try
			{
				long fee = _clients.get( clientId ).getMinimumFeeSync();
				unitySendMessage( "GetMinimumFeeSucceeded", callbackToJson( Long.toString( fee ), clientId ) );
			}
			catch( Exception e )
			{
				unitySendMessage( "GetMinimumFeeFailed", exceptionToJson( e, clientId ) );
				Log.e( TAG, "GetMinimumFee failed", e );
			}
		} ).start();
	}


	public void buildTransaction( final String accountId, final String toAddress, final String kinAmount, final int fee, final String memo )
	{
		new Thread( () -> {
			try
			{
				Log.i( TAG, "Preparing to build transaction. toAddress: " + toAddress + ", memo: " + memo );
				Transaction transaction;
				if( TextUtils.isEmpty( memo ) )
					transaction = _accounts.get( accountId ).buildTransactionSync( toAddress, new BigDecimal( kinAmount ), fee );
				else
					transaction = _accounts.get( accountId ).buildTransactionSync( toAddress, new BigDecimal( kinAmount ), fee, memo );

				_transactions.put( transaction.getId().id(), transaction );
				unitySendMessage("BuildTransactionSucceeded", transactionToJson( transaction, accountId ) );
			}
			catch( Exception e )
			{
				unitySendMessage( "BuildTransactionFailed", exceptionToJson( e, accountId ) );
				Log.e( TAG, "BuildTransaction failed", e );
			}
		} ).start();
	}


	public void sendTransaction( final String accountId, final String id )
	{
		new Thread( () -> {
			try
			{
				Transaction transaction = _transactions.remove( id );
				Log.i( TAG, "Preparing to send transaction: " + transaction.getDestination().getAccountId() );

				TransactionId transactionId = _accounts.get( accountId ).sendTransactionSync( transaction );
				unitySendMessage("SendTransactionSucceeded", callbackToJson( transactionId.id(), accountId ) );
			}
			catch( Exception e )
			{
				unitySendMessage( "SendTransactionFailed", exceptionToJson( e, accountId ) );
				Log.e( TAG, "SendTransaction failed", e );
			}
		} ).start();
	}


	public void sendWhitelistTransaction( final String accountId, final String id, final String whitelist )
	{
		new Thread( () -> {
			try
			{
				Transaction transaction = _transactions.remove( id );
				Log.i( TAG, "Preparing to send transaction: " + transaction.getDestination().getAccountId() );

				TransactionId transactionId = _accounts.get( accountId ).sendWhitelistTransactionSync( whitelist );
				unitySendMessage("SendTransactionSucceeded", callbackToJson( transactionId.id(), accountId ) );
			}
			catch( Exception e )
			{
				unitySendMessage( "SendTransactionFailed", exceptionToJson( e, accountId ) );
				Log.e( TAG, "SendTransaction failed", e );
			}
		} ).start();
	}


	public void addPaymentListener( final String accountId )
	{
		// we only need one listener on the native side. Multiple listeners can be added on the Unity side.
		if( _paymentListeners.containsKey( accountId ) )
			return;

		ListenerRegistration registration = _accounts.get( accountId ).addPaymentListener( paymentInfo -> unitySendMessage( "OnPayment", paymentInfoToJson( paymentInfo, accountId ) ) );
		_paymentListeners.put( accountId, registration );
	}


	public void removePaymentListener( String accountId )
	{
		if( _paymentListeners.containsKey( accountId ) )
		{
			_paymentListeners.get( accountId ).remove();
			_paymentListeners.remove( accountId );
		}
	}


	public void addBalanceListener( final String accountId )
	{
		// we only need one listener on the native side. Multiple listeners can be added on the Unity side.
		if( _balanceListeners.containsKey( accountId ) )
			return;

		ListenerRegistration registration = _accounts.get( accountId ).addBalanceListener( balance -> unitySendMessage( "OnBalance", callbackToJson( balance.value().toString(), accountId ) ) );
		_balanceListeners.put( accountId, registration );
	}


	public void removeBalanceListener( String accountId )
	{
		if( _balanceListeners.containsKey( accountId ) )
		{
			_balanceListeners.get( accountId ).remove();
			_balanceListeners.remove( accountId );
		}
	}


	public void addAccountCreationListener( final String accountId )
	{
		// we only need one listener on the native side. Multiple listeners can be added on the Unity side.
		if( _accountListeners.containsKey( accountId ) )
			return;

		ListenerRegistration registration = _accounts.get( accountId ).addAccountCreationListener( data -> unitySendMessage( "OnAccountCreated", accountId ) );
		_accountListeners.put( accountId, registration );
	}


	public void removeAccountCreationListener( String accountId )
	{
		if( _accountListeners.containsKey( accountId ) )
		{
			_accountListeners.get( accountId ).remove();
			_accountListeners.remove( accountId );
		}
	}

	//endregion

}
